jgrapht-core/src/main/java/org/jgrapht/alg/cycle/HierholzerEulerianCycle.java

```diff
@@ -0,0 +1,585 @@
+     * containing an <a href="http://mathworld.wolfram.com/EulerianCycle.html">Eulerian cycle</a>.
+     * 
+     * @param graph the input graph
+     * @return true if the graph is Eulerian, false otherwise
+     */
+    public boolean isEulerian(Graph<V, E> graph)
+    {
+        Objects.requireNonNull(graph, "Graph cannot be null");
+        if (graph.vertexSet().isEmpty()) {
+            // null-graph return false
+            return false;
+        } else if (graph.edgeSet().isEmpty()) {
+            // empty-graph with vertices
+            return true;
+        } else if (graph instanceof UndirectedGraph) {
+            UndirectedGraph<V, E> ug = TypeUtil.uncheckedCast(graph, null);
+            // check odd degrees
+            for (V v : ug.vertexSet()) {
+                if (ug.degreeOf(v) % 2 == 1) {
+                    return false;
+                }
+            }
+            // check that at most one connected component contains edges
+            boolean foundComponentWithEdges = false;
+            for (Set<V> component : new ConnectivityInspector<V, E>(ug).connectedSets()) {
+                for (V v : component) {
+                    if (ug.degreeOf(v) > 0) {
+                        if (foundComponentWithEdges) {
+                            return false;
+                        }
+                        foundComponentWithEdges = true;
+                        break;
+                    }
+                }
+            }
+            return true;
+        } else if (graph instanceof DirectedGraph) {
+            DirectedGraph<V, E> dg = TypeUtil.uncheckedCast(graph, null);
+            // check same in and out degrees
+            for (V v : dg.vertexSet()) {
+                if (dg.inDegreeOf(v) != dg.outDegreeOf(v)) {
+                    return false;
+                }
+            }
+            // check that at most one strongly connected component contains edges
+            boolean foundComponentWithEdges = false;
+            for (Set<V> component : new KosarajuStrongConnectivityInspector<V, E>(dg)
+                .stronglyConnectedSets())
+            {
+                for (V v : component) {
+                    if (dg.inDegreeOf(v) > 0 || dg.outDegreeOf(v) > 0) {
+                        if (foundComponentWithEdges) {
+                            return false;
+                        }
+                        foundComponentWithEdges = true;
+                        break;
+                    }
+                }
+            }
+            return true;
+        } else {
+            throw new IllegalArgumentException("Graph must be directed or undirected");
+        }
+    }
+
+    /**
+     * Compute an Eulerian cycle of a graph.
+     * 
+     * @param g the input graph
+     * @return an Eulerian cycle
+     * @throws IllegalArgumentException in case the graph is not Eulerian
+     */
+    public GraphPath<V, E> getEulerianCycle(Graph<V, E> g)
+    {
+        if (!GraphTests.isEulerian(g)) {
```

Comment:
Better to use this.isEulerian(). Imagine that someone changes the implementation of GraphTests.isEulerian(g) to:
return HierholzerEulerianCycle.getEulerianCycle(g) != null;

This would result in a cyclic definition.